### PR TITLE
fix(vendor): Strip excluded build targets 

### PR DESCRIPTION
**Pull request by epage**
### What does this PR try to resolve?

Fixes https://github.com/rust-lang/cargo/issues/14348

### How should we test and review this PR?

The commits are setup to be able to show, through the tests, how the `cargo package` and `cargo vendor` behavior diverge and then how `cargo vendor`s behavior changes over time.

This is a **very** hacky solution, duplicating the minimum of what
`prepare_for_publish` does to fix this one issue and in the least
intrusive way to the vendor code.

The intention is to keep this low risk for backporting to beta and
stable.
We need to revisit this, refactoring the `cargo package` code so that we
can call into that for each vendored dependency.

### CODE REVIEW

File: src/cargo/ops/vendor.rs [hunk 1]
```diff
@@ -392,6 +398,114 @@ fn cp_sources(
     Ok(())
 }
 
+fn prepare_for_vendor(
```
Comment (weihanglo) [inline src/cargo/ops/vendor.rs:401]: Even if we plan to revisit and refactor, a notice here could be useful for people about to diverge the logic to think twice.

```suggestion
/// This is a hack for resolving cargo/#14348.
/// It basically copies the normalization logic from `prepare_for_publish`.
/// TODO: need to revisit and consolidate both for `cargo package` and `cargo vendor`. 
fn prepare_for_vendor(
```

File: tests/testsuite/vendor.rs [hunk 2]
```diff
@@ -956,10 +1521,6 @@ fn git_deterministic() {
 # will likely look very different (and much more reasonable).
 # See Cargo.toml.orig for the original contents.
 
-bin = []
-test = []
-bench = []
```
Comment (weihanglo) [inline tests/testsuite/vendor.rs:961]: I am sorry that I didn't even notice this when reviewing that PR.

**Review by weihanglo — APPROVED** [review]

File: src/cargo/ops/vendor.rs (hunk 1 above)
Comment (weihanglo) [reply src/cargo/ops/vendor.rs:401]: `r=weihanglo` whenever you feel ready. The comment is not necessary.

File: tests/testsuite/vendor.rs (hunk 2 above)
Comment (epage) [reply tests/testsuite/vendor.rs:961]: Notice which?

**Comment by weihanglo** [comment]
I don't think it is controversial for beta backport. #14366 contains two commits to make the beta channel CI happy. @epage, are you working on the backport? I have some free time now and can handle it if needed.

**Comment by epage** [comment]
I can do the beta backport.

**Pull request merged**